text/0002-extracting-necko.md

```diff
@@ -0,0 +1,74 @@
+
+# Proposal
+
+Necko is a very feature-rich C++ networking library that is directly targeted at meeting the needs of single-user, single-main-process, single-profile, interactive applications built on a persistent profile directory in the traditional Mozilla model. It is not well-suited for use outside of the Gecko lifecycle — indeed, almost every aspect is designed for the Gecko world — and would require substantial effort in decoupling to be suited for embedded use. Even if so decoupled, it would still mandate the use of XPCOM and the NSPR.
+
+To extract Necko for reuse outside of Gecko, by non-XPCOM-based applications, would — almost by definition — require coming up with a plan to deal with XPCOM. That plan might look a lot like continuing historical efforts on deCOMtamination.
+
+Necko's data structures, interface definitions, error codes, macros, etc. *are* XPCOM's, and so eliminating XPCOM dependency from the module is likely to involve touching almost every line of code. [Automated tools to do so have been proposed](https://brendaneich.com/2006/10/mozilla-2/).
+
+(Going further to eliminate NSPR and mfbt data structures, yielding a more standard C++ library, would be further work still.)
+
+This is particularly relevant when we consider oxidation — the replacement of C++ and JS components of Firefox with Rust — and even more so when we consider the use of such components both outside and inside Firefox *from* Rust (rather than from existing non-Rust callers).
+
+There is ongoing work on landing Rust components in Firefox; most obstacles there are around XPCOM integration for Rust code, which currently requires writing lots of routine boilerplate by hand.
+
+There is also some initial scoping of work to extract and use Gecko C++ components from Servo, which is a similar problem to using Gecko C++ components from Rust background services or mobile applications. Those estimates are large (but not yet published), on the order of 3-5 engineer years for non-trivial components.
+
+Extracting such a module is not routine work: delicate balances need to be found between maintaining useful integration with Gecko (*e.g.*, Necko's `nsIChannel` permeates the codebase) versus designing for embedding, and between reimplementing Gecko dependencies in Rust for reuse versus reworking the component to avoid them.
+
+Our position is that extracting Necko from Gecko for reuse would be a significant amount of work. Doing so in a way that allows its use from ordinary Rust consumers, without shipping XPCOM/NSPR/*etc*., would be even more so.
+
+That amount of work is certainly much more than simply using the existing well-tested and idiomatic networking libraries available to Rust code. Given that the benefits of using Necko in its entirety (instead of *e.g.*, just using Hyper for HTTP) are relatively small, our proposal is that we **should not attempt to reuse Necko outside of Gecko**.
```

Comment:
Thanks for the feedback! I agree that we should avoid retrofitting the proposal to the final decision. The proposal is the assertion that initiates the research/analysis, and it's worth retaining it in its original form, even if the research/analysis winds up disproving it, so that readers understand the genesis of the process.

The final decision, likewise, might differ from the conclusion, so it's worth distinguishing between them as well. To provide an example from another domain: a drug company might propose regulatory approval of a drug (the proposal); an expert panel might then conclude that the drug should not be approved (the conclusion); and an administrator might then override that conclusion and approve the drug (the final decision).

In this case, the proposal was actually to "extract Necko," as I understand it (perhaps from discussions with @joewalker, although I don't remember exactly). @rnewman then did some research/analysis and drawn the conclusion that we should not do so. (It isn't clear that someone in a position of authority has made a "final decision" in this case, or even if it's necessary to do so.)

Reviewing the document, however, it isn't necessarily clear that "extract Necko" is the proposal and "we should not extract Gecko" is the conclusion, since the document starts with a Problem section that asks "Can we break out a Gecko component, such as Necko…?" and then continues with a Proposal section that states the conclusion "we should not attempt to reuse Necko outside of Gecko."

So I would revise it to start with a Proposal section that reads as a statement like "We should extract Necko for reuse in background services and on mobile platforms" and then rename the Proposal section to Conclusion. I'll submit a pull request to that effect, where we can discuss structural improvements further!